File: backend/app/shared/domain/networth.py
"""Pure, framework-free net-worth math. No I/O, no framework imports (mirrors
`shared/domain/budgeting`). The TS core in
`frontend/src/shared/domain/networth.ts` must agree on the same numbers.

Net worth is intentionally simple arithmetic, but it is a *formula* so it lives
here and is mirrored (per the repo's dual-core rule):

  net_worth  = assets_total - liabilities_total
  growth     = (current - previous) / |previous|   (None when previous == 0)
  allocation = each bucket's share of the positive total (fractions in [0, 1])
"""

from __future__ import annotations


def net_worth(assets_total: float, liabilities_total: float) -> float:
    return assets_total - liabilities_total


def growth(current: float, previous: float) -> float | None:
    """Fractional change from `previous` to `current`.

    None when there is no baseline (previous == 0) — a percentage change from
    zero is undefined rather than infinite.
    """
    if previous == 0:
        return None
    return (current - previous) / abs(previous)


def allocation(buckets: dict[str, float]) -> dict[str, float]:
    """Each positive bucket's share of the total. Non-positive buckets are 0."""
    total = sum(v for v in buckets.values() if v > 0)
    if total <= 0:
        return {k: 0.0 for k in buckets}
    return {k: (v / total if v > 0 else 0.0) for k, v in buckets.items()}

File: frontend/src/shared/domain/networth.ts
/**
 * Pure, framework-free net-worth math — the TS mirror of
 * `backend/app/shared/domain/networth.py`. Both cores must agree on the same
 * numbers (the mirrored tests assert identical values).
 */

export function netWorth(assetsTotal: number, liabilitiesTotal: number): number {
  return assetsTotal - liabilitiesTotal;
}

/** Fractional change from `previous` to `current`; null when previous is 0. */
export function growth(current: number, previous: number): number | null {
  if (previous === 0) return null;
  return (current - previous) / Math.abs(previous);
}

/** Each positive bucket's share of the total (fractions in [0, 1]). */
export function allocation(buckets: Record<string, number>): Record<string, number> {
  const total = Object.values(buckets).reduce((s, v) => (v > 0 ? s + v : s), 0);
  const out: Record<string, number> = {};
  for (const [k, v] of Object.entries(buckets)) {
    out[k] = total > 0 && v > 0 ? v / total : 0;
  }
  return out;
}
